### PR TITLE
split controller and apiserver start

### PR DESCRIPTION
**Pull request by deads2k**
This builds on @mfojtik's refactoring pull and separates the construction of the controller part of the process from the apiserver part of the process. More will be required, but this gets the informers under control and moves us in the correct direction.

@liggitt

### CODE REVIEW

**Comment by deads2k** [comment]
oh joy, conflict

**Comment by deads2k** [comment]
[test]

**Comment by deads2k** [comment]
@stevekuznetsov how hard would it be to have verify failures fail the job, but allow the unit tests to run anyway?

**Comment by stevekuznetsov** [comment]
Right now the job runs [two separate shell scripts](https://github.com/openshift/aos-cd-jobs/blob/master/sjb/config/test_cases/test_branch_origin_check.yml) for the verify and test steps, but if we squashed them together into one stage we could probably Bash our way around it.

**Comment by deads2k** [comment]
re[test]

File: pkg/cmd/server/kubernetes/master/controller/config.go [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+	// Remove the "normal" resource quota, because we run it as an openshift controller to cover all types
+	// TODO split openshift separately so we get upstream initialization here
+	delete(ret, "resourcequota")
+	// "serviceaccount-token" is used to create SA tokens for everyone else.  We special case this one.
+	delete(ret, "serviceaccount-token")
+
+	// TODO once the cloudProvider moves, move the configs out of here to where they need to be constructed
+	persistentVolumeController := PersistentVolumeControllerConfig{
+		RecyclerImage: c.RecyclerImage,
+		// TODO: In 3.7 this is renamed to 'Cloud' and is part of kubernetes ControllerContext
```
Comment (mfojtik) [inline pkg/cmd/server/kubernetes/master/controller/config.go]: guess we can remove this todos as you have todo in func godoc (which btw. should be in godoc format ;-)

File: pkg/cmd/server/origin/master_config.go [hunk 2]
```diff
@@ -415,6 +389,182 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+	if err != nil {
+		return nil, err
+	}
+	ret.DeployerControllerConfig = origincontrollers.DeployerControllerConfig{
+		ImageName:     imageTemplate.ExpandOrDie("deployer"),
+		Codec:         annotationCodec,
+		ClientEnvVars: vars,
+	}
+	ret.DeploymentConfigControllerConfig = origincontrollers.DeploymentConfigControllerConfig{
+		Codec: annotationCodec,
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master_config.go]: did we lost todo about moving codec to controller context? (maybe i never added that todo ;-)

File: pkg/cmd/server/origin/master_config.go [hunk 3]
```diff
@@ -415,6 +389,182 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+		HasBuilderEnabled: options.DisabledFeatures.Has(configapi.FeatureBuilder),
+		// TODO: make these consts in configapi
+		HasDeploymentsEnabled:  options.DisabledFeatures.Has("triggers.image.openshift.io/deployments"),
+		HasDaemonSetsEnabled:   options.DisabledFeatures.Has("triggers.image.openshift.io/daemonsets"),
+		HasStatefulSetsEnabled: options.DisabledFeatures.Has("triggers.image.openshift.io/statefulsets"),
+		HasCronJobsEnabled:     options.DisabledFeatures.Has("triggers.image.openshift.io/cronjobs"),
+	}
+	ret.ImageImportControllerOptions = origincontrollers.ImageImportControllerOptions{
+		MaxScheduledImageImportsPerMinute:          options.ImagePolicyConfig.MaxScheduledImageImportsPerMinute,
+		ResyncPeriod:                               10 * time.Minute,
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master_config.go]: wonder if we want to make the resyncPeriods configurable or we can just hardcode them in controllers

File: pkg/cmd/server/origin/master_config.go [hunk 4]
```diff
@@ -415,6 +389,182 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+
+	ret.ImageTriggerControllerConfig = origincontrollers.ImageTriggerControllerConfig{
+		HasBuilderEnabled: options.DisabledFeatures.Has(configapi.FeatureBuilder),
+		// TODO: make these consts in configapi
+		HasDeploymentsEnabled:  options.DisabledFeatures.Has("triggers.image.openshift.io/deployments"),
+		HasDaemonSetsEnabled:   options.DisabledFeatures.Has("triggers.image.openshift.io/daemonsets"),
+		HasStatefulSetsEnabled: options.DisabledFeatures.Has("triggers.image.openshift.io/statefulsets"),
+		HasCronJobsEnabled:     options.DisabledFeatures.Has("triggers.image.openshift.io/cronjobs"),
+	}
+	ret.ImageImportControllerOptions = origincontrollers.ImageImportControllerOptions{
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master_config.go]: s/ImageImportControllerOptions/ImageImportControllerConfig/ for consistency

File: pkg/cmd/server/origin/master_config.go [hunk 5]
```diff
@@ -415,6 +389,182 @@ func BuildMasterConfig(options configapi.MasterConfig) (*MasterConfig, error) {
...
+	ret.UnidlingControllerConfig = origincontrollers.UnidlingControllerConfig{
+		ResyncPeriod: 2 * time.Hour,
+	}
+	ret.IngressIPControllerConfig = origincontrollers.IngressIPControllerConfig{
+		IngressIPSyncPeriod:  10 * time.Minute,
+		IngressIPNetworkCIDR: options.NetworkConfig.IngressIPNetworkCIDR,
+	}
+
+	ret.OriginToRBACSyncControllerConfig = origincontrollers.OriginToRBACSyncControllerConfig{
+		PrivilegedRBACClient: kubeInternal.Rbac(),
```
Comment (mfojtik) [inline pkg/cmd/server/origin/master_config.go]: add comment about why this need privileged?

File: pkg/cmd/server/origin/run_components.go [hunk 6]
```diff
@@ -88,9 +92,21 @@ func (c *MasterConfig) RunProjectCache() {
 	go c.ProjectCache.Run(utilwait.NeverStop)
 }
 
+type SecurityAllocationController struct {
+	//c.Options.ProjectConfig.SecurityAllocator
+	SecurityAllocator *configapi.SecurityAllocator
+
+	//c.RESTOptionsGetter
```
Comment (mfojtik) [inline pkg/cmd/server/origin/run_components.go]: nuke this and above (i was doing the same when I was moving these ;-)

**Comment by mfojtik** [comment]
LGTM (with some nits)

also great stuff!

File: pkg/cmd/server/origin/master_config.go (hunk 2 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > did we lost todo about moving codec to controller context? (maybe i never added that todo ;-)

Having now gone through it, we don't want it generically available.  It turns out that it is being used improperly and will cause us versioning pain.  We should strive to eliminate it instead.

File: pkg/cmd/server/origin/master_config.go (hunk 3 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master_config.go]: > wonder if we want to make the resyncPeriods configurable or we can just hardcode them in controllers

Ultimately, they should be configurable.  This is a move of existing hardcodedness.

File: pkg/cmd/server/origin/master_config.go (hunk 2 above)
Comment (mfojtik) [inline pkg/cmd/server/origin/master_config.go]: fine with that

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
Comments were minor and addressed.  I plan to merge on green.

**Comment by deads2k** [comment]
@openshift/networking Can I get some help looking at the networking test here: https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_networking_minimal/3543 ?  I'm assuming my problem is `Reason:KubeletNotReady Message:runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized`, but I don't know where to go from there.

**Comment by deads2k** [comment]
[merge]

**Comment by deads2k** [comment]
[severity:blocker]

**Comment by dcbw** [comment]
> @openshift/networking Can I get some help looking at the networking test here: https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_extended_networking_minimal/3543 ? I'm assuming my problem is Reason:KubeletNotReady Message:runtime network not ready: NetworkReady=false reason:NetworkPluginNotReady message:docker: network plugin is not ready: cni config uninitialized, but I don't know where to go from there.

It means that openshift-sdn hasn't been able to get the ClusterNetwork object from the apiserver, or that it hasn't been able to read its own HostNetwork allocation from the master.  In both these cases, the plugin cannot initialized, which consists of writing a config file to /etc/cni/net.d which then kubelet looks for in a loop and when it finds one, magically updates node network readiness.

You should see log messages about the SDN reading the cluster network.  But I'm not sure how to get node messages out of AWS.  It should be fairly easy to reproduce locally though;  with your branch, can you just run:

hack/dind-cluster.sh start

and then if it fails the same way the extended test does, do:

docker exec -it openshift-node-2 bash

and then once that's dropped you into the node:

journalctl -b -u openshift-node > /tmp/node.log
exit
docker cp openshift-node-2:/tmp/node.log /tmp/node.log

and then somehow get node.log to one of us.

**Comment by danwinship** [comment]
> But I'm not sure how to get node messages out of AWS

Click on "S3 Artifacts" on the side of the test run, scroll down to near the bottom, click "scripts/networking-minimal/logs/multitenant/nettest-node-1/systemd.log.gz".

The nodes are failing with:

    node.go:325] error: SDN node startup failed: failed to get subnet for this host: nettest-node-1, error: timed out waiting for the condition

which means the master isn't correctly running the SDN controller. Actually, the master logs don't seem to show the openshift master running at all... the only "openshift-master"-related thing I see is:

    Jun 22 12:53:06 nettest-master systemd-journald[18]: Suppressed 3069 messages from /system.slice/openshift-master.service

The logs go another 2 minutes after that but have nothing from the openshift master. That seems really weird. If it had crashed, systemd should log something, so I guess it didn't crash, but maybe it deadlocked or something?

I'm going to try running this PR locally...

**Comment by danwinship** [comment]
oh, nm, apparently already fixed

**Comment by deads2k** [comment]
re[merge]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by deads2k** [comment]
re[test]

**Comment by smarterclayton** [comment]
Needs rebase looks like

**Comment by smarterclayton** [comment]
Flake, merged at head

**Pull request merged**